src/shogun/mathematics/graph/Graph_test.cc

```diff
@@ -0,0 +1,70 @@
+#include <gtest/gtest.h>
+
+#include "Graph.h"
+#include "nodes/Add.h"
+#include "nodes/Dot.h"
+#include "nodes/Input.h"
+#include "nodes/Multiply.h"
+#include "nodes/Sign.h"
+#include "nodes/Cast.h"
+#include "nodes/Equal.h"
+
+#include "test/GraphTest.h"
+
+using namespace shogun;
+using namespace shogun::graph;
+using namespace std;
+
+TYPED_TEST(GraphTest, perceptron_stochastic_gradient_descent)
+{
+	using NumericType = TypeParam;
+
+	if constexpr (!std::is_same_v<NumericType, float64_t>)
+		return;
+	else
+	{
+		size_t max_iters = 100; 
+		int64_t num_feat = 5;
+		int64_t num_examples = 1000;
+
+
+		auto X = make_shared<node::Input>(Shape{num_feat}, element_type::FLOAT64);
+		auto y = make_shared<node::Input>(Shape{}, element_type::FLOAT64);
+		auto w_input = make_shared<node::Input>(Shape{Shape::Dynamic}, element_type::FLOAT64);
+		auto b_input = make_shared<node::Input>(Shape{}, element_type::FLOAT64);
+	    
+	    auto predicted = make_shared<node::Dot>(X, w_input) + b_input;
+	    auto correct = make_shared<node::Equal>(make_shared<node::Sign>(predicted), y);
+	    auto gradient = make_shared<node::Cast>(correct, element_type::FLOAT64);
+	   	auto b = b_input + gradient;
+	    auto w = X * gradient;
+		 
+		auto graph = make_shared<Graph>(vector{X, y, w_input, b_input}, vector<shared_ptr<node::Node>>{w, b});
+		graph->build();
+
+		auto w_vector = SGVector<float64_t>(num_feat);
+		w_vector.set_const(NumericType{1} / num_feat);
+		NumericType b_value{0};
+		SGMatrix<NumericType> features(num_examples, num_feat);
+		SGVector<NumericType> labels(num_examples);
+
+		for (auto&& backend : this->m_backends)
+		{
+			if (backend != GRAPH_BACKEND::SHOGUN)
+				continue;
```

Comment:
@vigsterkr ngraph hangs when i run this, probably it is rebuilding the graph in each loop, because right now the graph is built during `execute` in the case of ngraph